lte/gateway/c/oai/tasks/amf/deregistration_request.cpp

```diff
@@ -0,0 +1,300 @@
+#include "amf_app_defs.h"
+#include "amf_recv.h"
+#include "amf_asDefs.h"
+#include "amf_as.h"
+#include "amf_sap.h"
+#include "amf_app_state_manager.h"
+
+using namespace std;
+namespace magma5g {
+amf_as_data_t amf_data_de_reg_sec;
+
+/*
+ * name : amf_handle_deregistration_ue_origin_req()
+ * Description: Starts processing de-registration request from UE.
+ *        Request comes from AS to AMF as UL NAS message.
+ *        Current scope is 3GPP connection, irrespective of
+ *        switch-off or normal de-registration.
+ *        re-registration required is out of mvc scope now.
+ */
+int amf_handle_deregistration_ue_origin_req(
+    amf_ue_ngap_id_t ue_id, DeRegistrationRequestUEInitMsg* msg, int amf_cause,
+    amf_nas_message_decode_status_t decode_status) {
+  OAILOG_FUNC_IN(LOG_NAS_AMF);
+  OAILOG_DEBUG(
+      LOG_NAS_AMF, "UE originated Deregistration procedures started\n");
+  int rc = RETURNerror;
+  amf_deregistration_request_ies_t params;
+  if (msg->m5gs_de_reg_type.switchoff) {
+    params.de_reg_type = AMF_SWITCHOFF_DEREGISTRATION;
+  } else {
+    params.de_reg_type = AMF_NORMAL_DEREGISTRATION;
+  }
+  /*value of access_type would be 1 or 2 or 3, 24-501 - 9.11.3.20 */
+  switch (msg->m5gs_de_reg_type.access_type) {
+    case 1:
+      params.de_reg_access_type = AMF_3GPP_ACCESS;
+      OAILOG_DEBUG(
+          LOG_NAS_AMF,
+          "Access type is AMF_3GPP_ACCESS for deregistration request from "
+          "UE\n");
+      break;
+    case 2:
+      params.de_reg_access_type = AMF_NONE_3GPP_ACCESS;
+      OAILOG_DEBUG(
+          LOG_NAS_AMF,
+          "Access type AMF_NONE_3GPP_ACCESS for deregistration request from "
+          "UE\n");
+      break;
+    case 3:
+      params.de_reg_access_type = AMF_3GPP_ACCESS_AND_NONE_3GPP_ACCESS;
+      OAILOG_DEBUG(
+          LOG_NAS_AMF,
+          "Access type AMF_3GPP_ACCESS_AND_NONE_3GPP_ACCESS for deregistration "
+          "request from UE\n");
+      break;
+    default:
+      OAILOG_DEBUG(
+          LOG_NAS_AMF, "Wrong access type received for deregistration\n");
+      OAILOG_FUNC_RETURN(LOG_NAS_AMF, rc);
+      break;
+  }
+  /*setting key set identifier as received from UE*/
+  params.ksi = msg->nas_key_set_identifier.nas_key_set_identifier;
+  increment_counter("ue_deregistration", 1, 1, "amf_cause", "ue_initiated");
+  rc = amf_proc_deregistration_request(ue_id, &params);
+  OAILOG_FUNC_RETURN(LOG_NAS_AMF, rc);
+}
+
+/*
+ * Function : amf_proc_deregistration_request
+ *
+ * Logic: Because it's UE originated, it has to be deleted from AMF
+ *        and other related NFs.
+ *        If switch off, no need to send deregistration accept in DL
+ *        Normal deregistration on 3GPP,
+ *        Send accept message in DL to UE.
+ *        Update UE state
+ *        Sends ue context release notification to NGAP
+ *        Clean up availble ue context
+ *        send message to SMF to remove all PDU sessions
+ *        AMF is not expecting response from SMF.
```

Comment:
Split this as comments inside the Function itself

Reply:
Resolved